### PR TITLE
Add AmbientContext and use it in LoggerPahseScope

### PR DESCRIPTION
**Pull request by vicancy**

### CODE REVIEW

File: src/Microsoft.DocAsCode.Common/AmbientContext.cs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+    using Newtonsoft.Json;
+
+    [Serializable]
+    public sealed class AmbientContext : SerializableContext, IDisposable
+    {
+        private string _contextId = string.Empty; // cache context id to enable lock free id retrieval.
+        private long _counter = 0;                // auto increment counter maintained during branch and trace entry creation.
+
+        private static readonly FieldInfo WrapperField = typeof(BitArray).GetField("_syncRoot", BindingFlags.Instance | BindingFlags.NonPublic);
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/AmbientContext.cs]: remove

File: src/Microsoft.DocAsCode.Common/AmbientContext.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static AmbientContext TryGetCurrentContext()
+        {
+            return CallContext.LogicalGetData(AMBCTX_NAME) as AmbientContext;
+        }
+
+        /// <summary>
+        /// Removes the current ambient context.
+        /// </summary>
+        public void Dispose()
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/AmbientContext.cs]: question:
using (var ac1 = AC.TryGetCurrentContext())
{
  // ac1
  using (var ac2 = ac.CreateBranch())
  {
    //ac2
  }
  // null?
}

File: src/Microsoft.DocAsCode.Common/AmbientContext.cs [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            return string.IsNullOrEmpty(id) ? new AmbientContext() : new AmbientContext(id);
+        }
+
+        /// <summary>
+        /// Generates the next id which correlates with other ids hierarchically.
+        /// </summary>
+        /// <returns>The generated correlation id.</returns>
+        public string GenerateNextCorrelationId()
+        {
+            return string.Format("{0}.{1}", Id, Interlocked.Increment(ref _counterRef[0]));
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/AmbientContext.cs]: >) [](start = 88, length = 1)

.ToString(), avoid boxing

File: src/Microsoft.DocAsCode.Common/Loggers/LoggerPhaseScope.cs [hunk 4]
```diff
@@ -10,7 +10,7 @@ public sealed class LoggerPhaseScope : IDisposable
     {
         private readonly string _originPhaseName;
         private readonly PerformanceScope _performanceScope;
-
+        private AmbientContext _ac;
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Common/Loggers/LoggerPhaseScope.cs]: readonly

File: src/Microsoft.DocAsCode.Common/AmbientContext.cs [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        private AmbientContext(string id)
+        {
+            Id = id;
+        }
+
+        private AmbientContext(AmbientContext context)
+        {
+            Id = context.Id;
+            _counter = 0;
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Common/AmbientContext.cs]: >_counter [](start = 12, length = 8)

why we need to set _counter = 0 here?

File: src/docfx/Program.cs [hunk 6]
```diff
@@ -98,6 +98,7 @@ internal static int ExecSubCommand(string[] args)
             finally
             {
                 scope?.Dispose();
+                AmbientContext.CurrentContext?.Dispose();
```
Comment (vwxyzh) [inline src/docfx/Program.cs]: remove?

**Comment by vwxyzh** [comment]
:shipit:

**Pull request merged**